### PR TITLE
change `NormalizesTo` to fully structurally normalize

### PR DESCRIPTION
**Pull request by lcnr**
notes in https://hackmd.io/wZ016dE4QKGIhrOnHLlThQ

need to also update the dev-guide once this PR lands. in short, the setup is now as follows:

`normalizes-to` internally implements one step normalization, applying that normalization to the `goal.predicate.term` causes the projected term to get recursively normalized. With this `normalizes-to` normalizes until the projected term is rigid, meaning that we normalize as many steps necessary, but at least 1.

To handle rigid aliases, we add another candidate only if the 1 to inf step normalization failed. With this `normalizes-to` is now full structural normalization. We can now change `AliasRelate` to simply emit `normalizes-to` goals for the rhs and lhs.

This avoids the concerns from https://github.com/rust-lang/trait-system-refactor-initiative/issues/103 and generally feels cleaner

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs [hunk 1]
```diff
@@ -30,14 +31,41 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
...
+                if alias.opt_kind(self.tcx()).is_some() {
+                    self.relate_rigid_alias_non_alias(
+                        param_env,
+                        alias,
+                        ty::Variance::Invariant,
+                        term,
+                    )?;
+                    self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                } else {
+                    // FIXME(ǵeneric_const_exprs): we currently do not support rigid
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:53]: ǵ -> g

**Comment by lcnr** [comment]
> TODO: this currently negatively affects perf because the structural norm is now nested alias-relate -> normalizes-to -> alias-relate. Intend to change `instantiate_normalizes_to_term` to use `normalizes_to` directly instead

Actually doesn't really negatively affect perf compared to the previous impl :thinking: trying the separate approach ends up causing worse perf by propagating more nested goals upwards, causing us to spend a lot more time in `try_evaluated_added_goals`

**Comment by BoxyUwU** [comment]
i mean @bors rollup=always (new solver)

**Pull request merged**